test/jdk/sun/security/validator/CertReplace.java

```diff
@@ -21,27 +21,186 @@
+
+        SecurityTools.keytool(ktBaseParameters +
+                              "-selfcert -alias ca");
+    }
+
+    /**
+     * This method creates certs for the Same DN test
+     *
+     * @throws Exception
+     */
+    private static void sameDn() throws Exception {
+
+        final String ktBaseParameters = "-storepass " + PASSWORD + " " +
+                                        "-keypass " + PASSWORD + " " +
+                                        "-keystore " + SAMEDN_JKS + " " +
+                                        "-keyalg rsa ";
+
+        final Path keystoreFilePath = Paths.get(SAMEDN_JKS);
+        Files.deleteIfExists(keystoreFilePath);
+
+        // 1. Generate 3 aliases in a keystore: ca1, ca2, user. The CAs' startdate
+        // is set to one year ago so that they are expired now
+        SecurityTools.keytool(ktBaseParameters +
+                              "-genkeypair -alias ca1 -dname CN=CA -keyalg rsa " +
+                              "-sigalg md5withrsa -ext bc -startdate -1y");
+        SecurityTools.keytool(ktBaseParameters +
+                              "-genkeypair -alias ca2 -dname CN=CA -keyalg rsa " +
+                              "-sigalg sha1withrsa -ext bc -startdate -1y");
+        SecurityTools.keytool(ktBaseParameters +
+                              "-genkeypair -alias user -dname CN=User -keyalg rsa");
+
+        final KeyStore keyStore = KeyStoreUtils.loadKeyStore(SAMEDN_JKS, PASSWORD);
+
+        // 2. Signing: ca -> user. The startdate is set to 1 minute in the past to ensure the certificate
+        // is valid at the time of validation and to prevent any issues with timing discrepancies
+        // Automatically saves the certs to the certs files
+
+        SecurityTools.keytool(ktBaseParameters +
+                              "-certreq -alias user -file user.req");
+        SecurityTools.keytool(ktBaseParameters +
+                              "-gencert -rfc -alias ca1 " +
+                              "-startdate -1M -infile user.req -outfile samedn1.certs");
+        SecurityTools.keytool(ktBaseParameters +
+                              "-gencert -rfc -alias ca2 " +
+                              "-startdate -1M -infile user.req -outfile samedn2.certs");
+
+        // 3. Remove user for cacerts
+        keyStore.deleteEntry("user");
+        keyStore.store(new FileOutputStream(CERTREPLACE_JKS), PASSWORD_CHAR_ARR);
+    }
+
     /**
      * @param args {cacerts keystore, cert chain}
      */
     public static void main(String[] args) throws Exception {
 
+        if (args[0].equals(CERTREPLACE_JKS)) {
+            certReplace();
+        } else if (args[0].equals(SAMEDN_JKS)) {
+            sameDn();
+        } else {
+            throw new RuntimeException("Not recognised test " + args[0]);
+        }
+
         KeyStore ks = KeyStore.getInstance("JKS");
         ks.load(new FileInputStream(args[0]), "changeit".toCharArray());
```

Comment:
You've defined `PASSWORD_CHAR_ARR` for this char array.

Also, close this `FileInputStream` to be safe, and the one in `createPath` method below.

Reply:
The opened input stream in the `createPath` method is still not closed.